spring-cloud-stream-modules-test-support/src/main/java/org/springframework/cloud/stream/modules/test/ftp/TestFtpServer.java

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright 2015 the original author or authors.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.stream.modules.test.ftp;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.Arrays;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+
+import org.apache.ftpserver.FtpServer;
+import org.apache.ftpserver.FtpServerFactory;
+import org.apache.ftpserver.ftplet.Authentication;
+import org.apache.ftpserver.ftplet.AuthenticationFailedException;
+import org.apache.ftpserver.ftplet.FtpException;
+import org.apache.ftpserver.ftplet.User;
+import org.apache.ftpserver.ftplet.UserManager;
+import org.apache.ftpserver.listener.ListenerFactory;
+import org.apache.ftpserver.usermanager.impl.BaseUser;
+import org.apache.ftpserver.usermanager.impl.ConcurrentLoginPermission;
+import org.apache.ftpserver.usermanager.impl.TransferRatePermission;
+import org.apache.ftpserver.usermanager.impl.WritePermission;
+import org.junit.rules.TemporaryFolder;
+
+import org.springframework.context.annotation.Bean;
+import org.springframework.integration.ftp.session.DefaultFtpSessionFactory;
+import org.springframework.integration.test.util.SocketUtils;
+
+/**
+ * Embedded FTP Server for test cases; exposes an associated session factory as a @Bean.
+ * 
+ * @author Artem Bilan
+ * @author Gary Russell
+ * @since 3.0
+ */
+public class TestFtpServer {
+
+	private final int ftpPort = SocketUtils.findAvailableServerSocket();
+
+	private final TemporaryFolder ftpFolder;
+
+	private final TemporaryFolder localFolder;
+
+	private volatile File ftpRootFolder;
+
+	private volatile File sourceFtpDirectory;
+
+	private volatile File targetFtpDirectory;
+
+	private volatile File sourceLocalDirectory;
+
+	private volatile File targetLocalDirectory;
+
+	private volatile FtpServer server;
+
+	public TestFtpServer(final String root) {
+		this.ftpFolder = new TemporaryFolder() {
+
+			@Override
+			public void create() throws IOException {
+				super.create();
+				ftpRootFolder = this.newFolder(root);
+				sourceFtpDirectory = new File(ftpRootFolder, "ftpSource");
+				sourceFtpDirectory.mkdir();
+				File file = new File(sourceFtpDirectory, "ftpSource1.txt");
```

Comment:
Sorry, I misread the location of this file. IMO, this should be made more generic (ie no default content) and moved to the common project, as it may be used to test several modules (eg ftp source *and* sink)